### PR TITLE
Apply new Playlist WebUI

### PR DESCRIPTION
**Pull request by sangwoo108**
Replace the experimental UI with new UI

* Added CSP rules to load our static resources like svg and font.

Resolves https://github.com/brave/brave-browser/issues/31536

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/browser/resources/utils/urlFixer.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { PlaylistItem } from 'gen/brave/components/playlist/common/mojom/playlist.mojom.m.js'
+import { Url } from 'gen/url/mojom/url.mojom.m'
+
+export function fixUpThumbnailURL (item: PlaylistItem) {
+  if (!item.thumbnailPath?.url) {
+    return
+  }
+
+  if (
+    item.thumbnailPath.url.startsWith('http://') ||
```
Comment (kdenhartog) [inline components/playlist/browser/resources/utils/urlFixer.ts:15]: Is it common for the thumbnails to be hosted as an http resource? It would be ideal to keep this as only `https://`

File: browser/ui/webui/playlist_ui.cc [hunk 2]
```diff
@@ -50,6 +50,12 @@ class UntrustedPlayerUI : public ui::UntrustedWebUIController {
         network::mojom::CSPDirectiveName::StyleSrc,
         std::string("style-src chrome-untrusted://resources "
                     "chrome-untrusted://brave-resources 'unsafe-inline';"));
+    source->OverrideContentSecurityPolicy(
+        network::mojom::CSPDirectiveName::ImgSrc,
+        std::string("img-src 'self' chrome-untrusted://resources;"));
```
Comment (kdenhartog) [inline browser/ui/webui/playlist_ui.cc:55]: Can you explain a bit further about the reasoning for utilizing `resources` here rather than the `playlist` frame? I'd be a bit concerned that storing these external resources even if they're sanitized under this origin because it's [used](https://github.com/search?q=repo%3Abrave%2Fbrave-core%20chrome-untrusted%3A%2F%2Fresources&type=code) by other parts of the browser in a way that likely doesn't assume that these are resources that weren't bundled with the browser. 

It would be ideal if we could isolate bundled resources and sanitized resources from each other under separate origins such as `chrome-untrusted://sanitized-resources` and keep `chrome-untrusted://resources` for only bundled resources.

**Review by kdenhartog — COMMENTED** [review]
Few comments, overall no major concerns with this approach but I want to make sure that we're not introducing inadvertent access to these resources in other parts of the browser like the wallet UI and the chat UI

File: components/playlist/browser/resources/utils/urlFixer.ts (hunk 1 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/utils/urlFixer.ts:15]: Not common but I think we need both, beacuse we can't sure what it could be in the wild 😄

File: browser/ui/webui/playlist_ui.cc (hunk 2 above)
Comment (sangwoo108) [reply browser/ui/webui/playlist_ui.cc:55]: Here, UntrustedPlayerUI doesn't load any external `images` as it only need some button icons. 

---

> I'd be a bit concerned that storing these external resources even if they're sanitized under this origin

The sanitized images are not served via `chrome-untrusted://resources`. They are served via `chrome-untrusted://playlist-data//playlist-data/{thumbnail/media}` and playlist data source will serve matching files from user profile directory. They are not bundled with our browser but stored in the directory on demand.

Comment (kdenhartog) [reply browser/ui/webui/playlist_ui.cc:55]: Awesome, thanks for clarifying. In that case, these all appear to be bundled resources so no concerns with this change to the CSP.

**Review by kdenhartog — APPROVED** [review]
Approving on behalf of the sec-team for the changes to the CSP here and accepting that we may need to support http based thumbnails still

File: components/playlist/browser/resources/components/playlistPlayer.tsx [hunk 3]
```diff
@@ -35,7 +37,10 @@ export default function PlaylistPlayer ({
           name={item.name}
           cached={item.cached}
           thumbnailUrl={item.thumbnailPath.url}
-          onClick={() => {}}
+          onClick={() => postMessageToPlayer({
+            actionType: types.PLAYLIST_ITEM_SELECTED,
+            data: { ...item, mediaPath: { url: item.mediaPath?.url } }
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/playlistPlayer.tsx:42]: It there some reason we can't do 
```suggestion
            data: { ...item, mediaPath: item.mediaPath }
```
here? Do we want to make sure we only send the url bit?

File: components/playlist/browser/resources/components/playlistsCatalog.tsx [hunk 4]
```diff
@@ -120,12 +120,10 @@ function PlaylistThumbnail (props: ThumbnailProps) {
 // A card UI for representing a playlist.
 function PlaylistCard ({ playlist }: { playlist: Playlist }) {
   const isDefaultPlaylist = playlist.id === 'default'
-
   const thumbnailUrl = React.useMemo(() => {
     return playlist.items?.find(item => item.thumbnailPath?.url)?.thumbnailPath
-      .url
+      ?.url
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/playlistsCatalog.tsx:125]: Is this change needed? I think the thumbnail path should always be truthy if we find a match.

File: components/playlist/browser/resources/playlist.tsx [hunk 5]
```diff
@@ -22,18 +27,19 @@ import store from './store'
...
-        dark={DarkTheme}
-        light={Theme}
-      >
-        <App />
-      </BraveCoreThemeProvider>
-    </Provider>,
-    document.getElementById('root'))
+    <BrowserRouter>
+      <Provider store={store}>
+        <BraveCoreThemeProvider dark={DarkTheme} light={Theme}>
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/playlist.tsx:32]: Does this version of playlist (v1) actually support theming? If not you *might* be able to remove this :laughing: Leo theming doesn't use this at all.

File: components/playlist/browser/resources/reducers/playlist_reducer.ts [hunk 6]
```diff
@@ -20,10 +22,21 @@ const playlistReducer: Reducer<PlaylistData | undefined> = (
...
+            fixUpMediaURL(item)
+          }
+        }
+      }
+
+      let currentList: Playlist | undefined
       if (state.currentList) {
-        currentList = playlists.find((list: PlaylistMojo.Playlist) => {
+        currentList = playlists.find((list: Playlist) => {
           return list.id === state?.currentList?.id
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/reducers/playlist_reducer.ts:40]: State can never be undefined, I think, because of line 20
```suggestion
          return list.id === state.currentList?.id
```

File: ui/webui/resources/BUILD.gn [hunk 7]
```diff
@@ -148,12 +148,14 @@ leo_icons = [
   "arrow-right.svg",
   "arrow-small-down.svg",
   "arrow-small-up.svg",
+  "autoplay-off.svg",
```
Comment (fallaciousreasoning) [inline ui/webui/resources/BUILD.gn:151]: Thanks for sorting these!

**Review by fallaciousreasoning — APPROVED** [review]
LGTM with some (optional) nits

File: components/playlist/browser/resources/reducers/playlist_reducer.ts (hunk 6 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/reducers/playlist_reducer.ts:40]: Yeah, I agree. Unfortunately, my local dev env seems to complain it could be undefined, which is a false alarm.

File: components/playlist/browser/resources/playlist.tsx (hunk 5 above)
Comment (sangwoo108) [reply components/playlist/browser/resources/playlist.tsx]: Yeah, that's true. I don't think I've used it at all 😄

File: components/playlist/browser/resources/components/playlistPlayer.tsx [hunk 8]
```diff
@@ -37,10 +37,12 @@ export default function PlaylistPlayer ({
...
           cached={item.cached}
           thumbnailUrl={item.thumbnailPath.url}
-          onClick={() => postMessageToPlayer({
-            actionType: types.PLAYLIST_ITEM_SELECTED,
-            data: { ...item, mediaPath: { url: item.mediaPath?.url } }
-          })}
+          onClick={() =>
+            postMessageToPlayer({
+              actionType: types.PLAYLIST_ITEM_SELECTED,
+              data: item
```
Comment (fallaciousreasoning) [inline components/playlist/browser/resources/components/playlistPlayer.tsx:43]: :+1: even better!

**Review by fallaciousreasoning — APPROVED** [review]
LGTM!

File: components/playlist/browser/resources/reducers/playlist_reducer.ts (hunk 6 above)
Comment (fallaciousreasoning) [reply components/playlist/browser/resources/reducers/playlist_reducer.ts:40]: Ah right, the makes sense :laughing: sorry for the noise!
